### PR TITLE
planner: fix incorrectly using the schema for plan cache

### LINKED ISSUE #56733: data inconsistency in table after injection one of tikv network partition during adding index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、enable tidb_enable_dist_task and global sort
2、run sysbench
3、add index for one table
4、injection one of tikv network partition

logs:
[logs.zip](https://github.com/user-attachments/files/17554343/logs.zip)

### 2. What did you expect to see? (Required)
admin check success

### 3. What did you see instead (Required)
admin check failed
admin check failed (Error 8223 (HY000): data inconsistency in table: sbtest1, index: index_test_1729177017422, handle: 34904611, index-values:"" != record-values:"handle: 34904611, values: [KindString 22819232683-65590422292-63271134372-34697476090-61344961071-33702946624-01334278487-33665526949-83071148098-02089169179]")
operatorLogs:
[2024-10-17 22:56:57] ###### start adding index
ALTER TABLE `sbtest1` ADD INDEX `index_test_1729177017422`(`c`)
[2024-10-17 22:56:57] ###### wait for ddl job finish
[2024-10-17 23:29:24] ###### ddl job finished
select job_id, db_name, table_name, job_type, create_time, start_time, end_time, state, query from information_schema.ddl_jobs where query = 'ALTER TABLE `sbtest1` ADD INDEX `index_test_1729177017422`(`c`)'
jobId: 491, job type: add index /* ingest cloud */, state: synced
add index done, it takes: 32m27.423231672s
[2024-10-17 23:29:24] ###### start admin check
admin check index sbtest1 index_test_1729177017422

### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v8.4.0
Edition: Community
Git Commit Hash: ff1206c7799cf9c0cf422498bc9b3dc8a5d44013
Git Branch: HEAD
UTC Build Time: 2024-10-17 01:59:30
GoVersion: go1.23.2
Race Enabled: false
Check Table Before Drop: false
Store: unistore
2024-10-17T22:56:48.981+0800

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #56733

Problem Summary:
If the table ID is changed for a given table name, the plan cache key is not updated.
For an insert statement, it may use the outdated schema, resulting in incorrect data or query results.

### What changed and how does it work?
Make sure the `tbl` is updated in function planCachePreprocess() if tableByID doesn't return the table info.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
fix incorrectly using the schema for plan cache
```

### CODE REVIEW

File: pkg/planner/core/plan_cache.go [hunk 1]
```diff
@@ -116,10 +116,11 @@ func planCachePreprocess(ctx context.Context, sctx sessionctx.Context, isNonPrep
 				return plannererrors.ErrSchemaChanged.GenWithStack("Schema change caused error: %s", err.Error())
 			}
 			delete(stmt.RelateVersion, stmt.tbls[i].Meta().ID)
+			tbl = tblByName
 			stmt.tbls[i] = tblByName
 			stmt.RelateVersion[tblByName.Meta().ID] = tblByName.Meta().Revision
 		}
-		newTbl, err := tryLockMDLAndUpdateSchemaIfNecessary(ctx, sctx.GetPlanCtx(), stmt.dbName[i], stmt.tbls[i], is)
+		newTbl, err := tryLockMDLAndUpdateSchemaIfNecessary(ctx, sctx.GetPlanCtx(), stmt.dbName[i], tbl, is)
```
Comment (zimulala) [inline pkg/planner/core/plan_cache.go:123]: Could we add some comments for this change(using `tbl`)?

File: pkg/planner/core/plan_cache.go [hunk 2]
```diff
@@ -129,7 +130,7 @@ func planCachePreprocess(ctx context.Context, sctx sessionctx.Context, isNonPrep
 		// The version of stmt.tbls[i] is taken from the prepare statement and is revision v1.
 		// When stmt.tbls[i] is locked in MDL, the revision of newTbl is also v1.
 		// The revision of tbl is v2. The reason may have other statements trigger "tryLockMDLAndUpdateSchemaIfNecessary" before, leading to tbl revision update.
-		if stmt.tbls[i].Meta().Revision != newTbl.Meta().Revision || (tbl != nil && tbl.Meta().Revision != newTbl.Meta().Revision) {
+		if stmt.tbls[i].Meta().Revision != newTbl.Meta().Revision {
```
Comment (zimulala) [inline pkg/planner/core/plan_cache.go:133]: Please remove line128-132.

File: pkg/ddl/tests/metadatalock/mdl_test.go [hunk 3]
```diff
@@ -997,6 +1001,104 @@ func TestMDLPreparePlanCacheExecute2(t *testing.T) {
 	tk.MustExec("admin check table t")
 }
 
+func TestMDLPreparePlanCacheExecuteInsert(t *testing.T) {
```
Comment (zimulala) [inline pkg/ddl/tests/metadatalock/mdl_test.go:1004]: This test is quite complicated, could you add some comments?

File: pkg/planner/core/plan_cache.go (hunk 1 above)
Comment (wjhuang2016) [reply pkg/planner/core/plan_cache.go:123]: tbl is natural I think, however, stmt.tbls[i] is strange here.

File: pkg/ddl/tests/metadatalock/mdl_test.go (hunk 3 above)
Comment (wjhuang2016) [reply pkg/ddl/tests/metadatalock/mdl_test.go:1004]: okey

**Review by tangenta — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**